src/login/logind.c

```diff
@@ -411,6 +412,43 @@ static int manager_enumerate_users(Manager *m) {
         return r;
 }
 
+static int parse_fdname(const char *fdname, char **id, dev_t *dev)
+{
+        _cleanup_strv_free_ char **parts = NULL;
+        unsigned int major, minor;
+        char *session_id;
+        int r;
+
+        parts = strv_split(fdname, "-");
+        if (!parts)
+                return -ENOMEM;
+        if (strv_length(parts) < 5)
+                return -EINVAL;
+
+        if (!streq(parts[0], "device"))
+                return -EINVAL;
+
+        r = safe_atou(parts[1], &major);
+        if (r < 0)
+                return r;
+        r = safe_atou(parts[2], &minor);
+        if (r < 0)
+                return r;
+
+        if (!streq(parts[3], "session"))
+                return -EINVAL;
```

Comment:
This looks like a sscanf would be easier than doing the parsing by hand.

Reply:
Thanks for your review!  I'm thinking of sticking with the "by hand" approach here.

%s is greedy and consumes all non-whitespace, so it would not work after changing to `session-%s-device-%d-%d`.  *scanf() also makes people nervous due to the insecure design of %s.  With sscanf() you can work around it but it's still hacky.

I since noticed

> The session id was put at the end, just in case it contained a separator

was not necessarily needed, so the fun detail with strjoin() below can go away.